// aminul-new-fca/src/forwardMessage.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function forwardMessage(messageID, targetThreadIDs, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!messageID) {
      return callback({ error: "messageID is required" });
    }

    if (!targetThreadIDs) {
      return callback({ error: "targetThreadIDs is required" });
    }

    if (!Array.isArray(targetThreadIDs)) {
      targetThreadIDs = [targetThreadIDs];
    }

    if (targetThreadIDs.length === 0) {
      return callback({ error: "At least one target thread is required" });
    }

    var form = {
      message_id: messageID,
      forward_message_id: messageID
    };

    targetThreadIDs.forEach(function(threadID, index) {
      form["thread_fbid[" + index + "]"] = threadID;
    });

    defaultFuncs
      .post("https://www.facebook.com/ajax/mercury/forward_messages.php", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        callback(null, {
          messageID: messageID,
          forwardedTo: targetThreadIDs,
          success: true
        });
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
