// aminul-new-fca/src/getMutualFriends.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function getMutualFriends(userID, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!userID) {
      return callback({ error: "userID is required" });
    }

    var form = {
      uid: userID,
      viewer: ctx.userID,
      source_ref: "FriendsBrowser",
      ref: "profile_mutual"
    };

    defaultFuncs
      .post("https://www.facebook.com/ajax/browser/list/mutualfriends/", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        var mutualFriends = [];
        
        if (resData.payload && resData.payload.entries) {
          resData.payload.entries.forEach(function(entry) {
            mutualFriends.push({
              userID: entry.uid.toString(),
              name: entry.title,
              profileUrl: entry.path,
              thumbSrc: entry.photo
            });
          });
        }

        callback(null, {
          count: mutualFriends.length,
          friends: mutualFriends
        });
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
