// aminul-new-fca/src/getOnlineUsers.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function getOnlineUsers(callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    var form = {
      viewer: ctx.userID,
      __user: ctx.userID
    };

    defaultFuncs
      .post("https://www.facebook.com/ajax/chat/buddy_list.php", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        var onlineUsers = [];
        
        if (resData.payload && resData.payload.buddy_list) {
          var buddyList = resData.payload.buddy_list;
          
          Object.keys(buddyList).forEach(function(userID) {
            var user = buddyList[userID];
            if (user.la) {
              onlineUsers.push({
                userID: userID,
                lastActive: user.la * 1000,
                status: user.a || 0,
                isIdle: user.a === 2
              });
            }
          });
        }

        callback(null, onlineUsers);
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
