// aminul-new-fca/src/getRecentMessages.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function getRecentMessages(threadID, limit, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof limit === "function") {
      callback = limit;
      limit = 20;
    }

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!threadID) {
      return callback({ error: "threadID is required" });
    }

    limit = Math.min(limit || 20, 100);

    api.getThreadHistory(threadID, limit, undefined, function(err, history) {
      if (err) {
        return callback(err);
      }

      var messages = history.map(function(msg) {
        return {
          messageID: msg.messageID,
          threadID: msg.threadID,
          senderID: msg.senderID,
          body: msg.body || "",
          timestamp: msg.timestamp,
          attachments: msg.attachments || [],
          isUnread: msg.isUnread || false,
          type: msg.type
        };
      });

      callback(null, {
        threadID: threadID,
        count: messages.length,
        messages: messages
      });
    });

    return returnPromise;
  };
};
