// aminul-new-fca/src/getThreadMembers.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function getThreadMembers(threadID, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!threadID) {
      return callback({ error: "threadID is required" });
    }

    api.getThreadInfo(threadID, function(err, threadInfo) {
      if (err) {
        return callback(err);
      }

      if (!threadInfo) {
        return callback({ error: "Thread not found" });
      }

      var members = [];
      var participantIDs = threadInfo.participantIDs || [];
      var nicknames = threadInfo.nicknames || {};
      var adminIDs = threadInfo.adminIDs || [];

      var userInfoPromises = participantIDs.map(function(userID) {
        return new Promise(function(resolve) {
          api.getUserInfo(userID, function(err, info) {
            if (err || !info || !info[userID]) {
              resolve({
                userID: userID,
                name: "Unknown User",
                nickname: nicknames[userID] || null,
                isAdmin: adminIDs.includes(userID),
                error: true
              });
            } else {
              var user = info[userID];
              resolve({
                userID: userID,
                name: user.name,
                firstName: user.firstName,
                nickname: nicknames[userID] || null,
                thumbSrc: user.thumbSrc,
                profileUrl: user.profileUrl,
                gender: user.gender,
                isAdmin: adminIDs.some(function(admin) {
                  return admin.id === userID || admin === userID;
                }),
                isFriend: user.isFriend || false
              });
            }
          });
        });
      });

      Promise.all(userInfoPromises)
        .then(function(memberResults) {
          callback(null, {
            threadID: threadID,
            threadName: threadInfo.threadName,
            isGroup: threadInfo.isGroup,
            memberCount: memberResults.length,
            members: memberResults
          });
        })
        .catch(function(err) {
          callback(err);
        });
    });

    return returnPromise;
  };
};
