// aminul-new-fca/src/getThreadParticipants.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function getThreadParticipants(threadID, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!threadID) {
      return callback({ error: "threadID is required" });
    }

    var form = {
      thread_ids: [threadID],
      "thread_fbids[0]": threadID
    };

    defaultFuncs
      .post("https://www.facebook.com/ajax/mercury/participants_info.php", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        var participants = [];
        
        if (resData.payload && resData.payload.profiles) {
          var profiles = resData.payload.profiles;
          
          Object.keys(profiles).forEach(function(userID) {
            var profile = profiles[userID];
            participants.push({
              userID: userID,
              name: profile.name,
              firstName: profile.firstName,
              vanity: profile.vanity || null,
              thumbSrc: profile.thumbSrc,
              profileUrl: profile.uri,
              gender: profile.gender,
              type: profile.type,
              isFriend: profile.is_friend || false
            });
          });
        }

        callback(null, participants);
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
