// aminul-new-fca/src/getUnreadThreads.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function getUnreadThreads(limit, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof limit === "function") {
      callback = limit;
      limit = 20;
    }

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    limit = limit || 20;

    var form = {
      "folders[0]": "inbox",
      limit: limit,
      offset: 0,
      filter: "unread"
    };

    defaultFuncs
      .post("https://www.facebook.com/ajax/mercury/unread_threads.php", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        var unreadThreads = [];
        
        if (resData.payload && resData.payload.threads) {
          resData.payload.threads.forEach(function(thread) {
            unreadThreads.push({
              threadID: thread.thread_fbid || thread.other_user_fbid,
              name: thread.name,
              unreadCount: thread.unread_count || 1,
              snippet: thread.snippet,
              timestamp: thread.timestamp,
              isGroup: !!thread.thread_fbid
            });
          });
        }

        callback(null, unreadThreads);
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
