// aminul-new-fca/src/searchMessages.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function searchMessages(query, threadID, limit, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof threadID === "function") {
      callback = threadID;
      threadID = null;
      limit = 20;
    } else if (typeof limit === "function") {
      callback = limit;
      limit = 20;
    }

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!query || query.trim().length === 0) {
      return callback({ error: "query is required" });
    }

    limit = Math.min(limit || 20, 50);

    var form = {
      query: query.trim(),
      limit: limit,
      offset: 0
    };

    if (threadID) {
      form.thread_fbid = threadID;
    }

    defaultFuncs
      .post("https://www.facebook.com/ajax/mercury/search_messages.php", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        var results = [];
        
        if (resData.payload && resData.payload.search_snippets) {
          Object.keys(resData.payload.search_snippets).forEach(function(msgID) {
            var snippet = resData.payload.search_snippets[msgID];
            results.push({
              messageID: msgID,
              threadID: snippet.thread_fbid || snippet.other_user_fbid,
              snippet: snippet.snippet,
              timestamp: snippet.timestamp,
              senderID: snippet.author
            });
          });
        }

        callback(null, {
          query: query,
          count: results.length,
          results: results
        });
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
