// aminul-new-fca/src/setTypingStatus.js
"use strict";

var utils = require("../utils");

module.exports = function(defaultFuncs, api, ctx) {
  return function setTypingStatus(threadID, isTyping, callback) {
    var resolveFunc = function() {};
    var rejectFunc = function() {};
    var returnPromise = new Promise(function(resolve, reject) {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (typeof isTyping === "function") {
      callback = isTyping;
      isTyping = true;
    }

    if (typeof callback !== "function") {
      callback = function(err, data) {
        if (err) return rejectFunc(err);
        resolveFunc(data);
      };
    }

    if (!threadID) {
      return callback({ error: "threadID is required" });
    }

    isTyping = isTyping !== false;

    var form = {
      source: "mercury-chat",
      thread: threadID,
      typ: isTyping ? 1 : 0
    };

    defaultFuncs
      .post("https://www.facebook.com/ajax/messaging/typ.php", ctx.jar, form)
      .then(utils.parseAndCheckLogin(ctx, defaultFuncs))
      .then(function(resData) {
        if (resData.error) {
          throw resData;
        }

        callback(null, { 
          threadID: threadID, 
          isTyping: isTyping,
          success: true 
        });
      })
      .catch(function(err) {
        callback(err);
      });

    return returnPromise;
  };
};
